gear-programs/bridging-payment/src/services/mod.rs

```diff
@@ -0,0 +1,193 @@
+        unsafe {
+            DATA.as_ref()
+                .expect("BridgePaymentData::seed() should be called")
+        }
+    }
+
+    fn data_mut(&mut self) -> &mut BridgePaymentData {
+        unsafe {
+            DATA.as_mut()
+                .expect("BridgePaymentData::seed() should be called")
+        }
+    }
+
+    fn config(&self) -> &Config {
+        unsafe {
+            CONFIG
+                .as_ref()
+                .expect("BridgePaymentData::seed() should be called")
+        }
+    }
+
+    fn config_mut(&self) -> &mut Config {
+        unsafe {
+            CONFIG
+                .as_mut()
+                .expect("BridgePaymentData::seed() should be called")
+        }
+    }
+
+    pub fn handle_reply() {}
+}
+
+#[gservice(events = BridgePaymentEvents)]
+impl<T> BridgePayment<T>
+where
+    T: ExecContext,
+{
+    pub fn set_fee(&mut self, fee: u128) {
+        let data = self.data();
+        if data.admin_address != self.exec_context.actor_id() {
+            panic!("Not admin");
+        }
+        let config: &mut Config = self.config_mut();
+        config.fee = fee;
+    }
+
+    pub fn reclaim_fee(&mut self) {
+        let data = self.data();
+        if data.admin_address != self.exec_context.actor_id() {
+            panic!("Not admin");
+        }
+        let fee_balance = exec::value_available();
+        msg::send_with_gas(data.admin_address.into(), "", 0, fee_balance)
+            .expect("Failed to reclaim fees");
+    }
+
+    pub async fn request_to_gateway(
+        &mut self,
+        amount: U256,
+        receiver: H160,
+        vara_token_id: ActorId,
+    ) {
+        let data = self.data();
+        let config = self.config();
+
+        let attached_value = msg::value();
+
+        if attached_value < config.fee {
+            panic!("Not enough fee");
+        }
+        let sender = self.exec_context.actor_id();
+        match utils::send_message_to_gateway(
+            data.grc20_gateway_address,
+            sender,
+            vara_token_id,
+            amount,
+            receiver,
+            config,
+        )
+        .await
+        {
+            Ok(Ok((nonce, eth_token_id))) => {
+                self.notify_on(BridgePaymentEvents::TeleportVaraToEth {
+                    nonce,
+                    sender,
+                    amount,
+                    receiver: receiver.into(),
+                    eth_token_id,
+                })
+                .expect("Error in depositing event");
+
+                // return remainder
+                let refund = attached_value - self.config().fee;
+                if refund >= exec::env_vars().existential_deposit {
+                    handle_refund(self.exec_context.actor_id(), refund);
+                }
+            }
+
+            _ => {
+                // Handle refund for the full attached value on any error
+                handle_refund(self.exec_context.actor_id(), attached_value);
```

Comment:
Why not just to panic here?

Reply:
Because `.await` has already been made and the sent funds are saved in the contract state. Panic won't bring them back